### PR TITLE
ESQL: improve mv_expand planning

### PR DESCRIPTION
**Pull request by astefan**
This is a collection of improvements made to how mv_expand is planned.
There are two mains areas in terms of optimization rules:
- `limit` after `mv_expand` can be duplicated before `mv_expand` when there is no `sort` after `mv_expand`. Here I used [@luigidellaquila's approach](https://github.com/elastic/elasticsearch/pull/102545/files#diff-fda161048777823e261492345cbeefb93f4afd4d583cfce41a94069459507466) to link a `limit` to a `mv_expand` as a way of not copying multiple times the same limit past mv_expand. The edge case query that really needs this is [this query](https://github.com/elastic/elasticsearch/pull/103582/files#diff-427544eaca2f06adb1b5850278a22573fb696564260a34a8f5580ff66bbbd95aR1147) `from test | where false | mv_expand first_name`.
- mv_expand can be pushed down past by sort in certain situations:
  - every time there is a `sort` after `mv_expand` the `mv_expand` can be pushed down
  - when there is no `sort` after `mv_expand` and the `limit` cannot be duplicated past `mv_expand`
 
While trying out different queries there is one where the two rules above do not apply:

```
from test
            | sort last_name
            | keep first_name
            | mv_expand first_name
            | where first_name is not null
```
This query is present here in tests https://github.com/elastic/elasticsearch/pull/103582/files#diff-427544eaca2f06adb1b5850278a22573fb696564260a34a8f5580ff66bbbd95aR1058.

Also, while testing I realized we can push down `mv_expand` past `keep`, but we have a faulty handling of `mv_expand`. Reported [here](https://github.com/elastic/elasticsearch/issues/103543).

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 1]
```diff
@@ -542,31 +517,97 @@ private static MvExpand descendantMvExpand(UnaryPlan unary) {
...
+                    unary = unaryPlan;
                 } else {
                     break;
                 }
             }
-            return null;
+
+            if (mvExpand != null) {
+                // if the filter is on the expanded values and the order of the results doesn't change, push down mv_expand past sort
+                if (isFiltered && hasOrderBy == false) {
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:579]: I think `mvExpand == null` and `hasOrderBy` are mutually exclusive. If so, maybe this can be written a bit easier.

**Review by alex-spies — COMMENTED** [review]
I think the general approach is good. I didn't go through all the test cases yet, but otherwise made some comments.

I think as our optimizer grows and rules become more complex, it'd be useful to have unit tests for individual rules which test only the given, single rule (in contrast to testing the whole optimizer on given queries). I find it difficult to verify if the rules will produce equivalent plans in all situations.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 2]
```diff
@@ -1035,24 +1134,73 @@ public void testPushDownLimitThroughMultipleSort_AfterMvExpand() {
...
+        mvExpand = as(limit.child(), MvExpand.class);
+        limit = as(mvExpand.child(), Limit.class);
+        assertThat(limit.limit().fold(), equalTo(7300));
+        as(limit.child(), Row.class);
+    }
+
+    /**
+     * Expected
+     * EsqlProject[[emp_no{f}#10, first_name{r}#20, salary{f}#15]]
+     * \_TopN[[Order[first_name{r}#20,ASC,LAST]],5[INTEGER]]
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:1194]: I guess two consecutive TopNs could be fused. (Out of scope for this PR)

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 3]
```diff
@@ -1162,22 +1416,64 @@ public void testPushDownLimit_PastEvalAndMvExpand() {
...
+     * Expected
+     * TopN[[Order[salary{f}#11,ASC,LAST]],5[INTEGER]]
+     * \_MvExpand[first_name{f}#7,first_name{r}#16]
+     *   \_EsRelation[test][_meta_field{f}#12, emp_no{f}#6, first_name{f}#7, ge..]
+     */
+    public void testPushDownMvExpand_PastSameFieldSorting2() {
+        LogicalPlan plan = optimizedPlan("""
+            from test
+            | sort first_name
+            | mv_expand first_name
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:1432]: I think here we could duplicate the topn and put it before the MV expand.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 4]
```diff
@@ -1105,6 +1314,52 @@ public void testDontPushDownLimitPastAggregate_AndMvExpand() {
...
+    /**
+     * Expected
+     * TopN[[Order[emp_no{f}#6,ASC,LAST]],5[INTEGER]]
+     * \_Filter[ISNOTNULL(first_name{r}#16)]
+     *   \_MvExpand[first_name{f}#7,first_name{r}#16]
+     *     \_EsRelation[test][_meta_field{f}#12, emp_no{f}#6, first_name{f}#7, ge..]
+     */
+    public void testDontPushDownLimit_ButPushMvExpandPastSort2() {
+        LogicalPlan plan = optimizedPlan("""
+            from test
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:1349]: Looks like this is a duplicate of the previous test.

**Review by alex-spies — COMMENTED** [review]
Alright, had another go. Left some observations, and overall still agree with the approach :)

It's great that you agreed so many tests. This must have been fun, getting all the edge cases right. Thanks Andrei!

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java (hunk 2 above)
Comment (alex-spies) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:1194]: Ugh, I was clearly too tired when writing this, please disregard.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 5]
```diff
@@ -481,59 +482,33 @@ private static Limit descendantLimit(UnaryPlan unary) {
         }
     }
 
-    static class DuplicateLimitAfterMvExpand extends OptimizerRules.OptimizerRule<Limit> {
+    private static boolean isTransparentForLimit(LogicalPlan plan) {
+        return plan instanceof Eval || plan instanceof Project || plan instanceof RegexExtract || plan instanceof Enrich;
+    }
 
+    protected static class PushDownMvExpandPastOrderBy extends OptimizerRules.OptimizerRule<OrderBy> {
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:489]: I tried to reconstruct the behavior of this rule and the reasoning behind it, and I have to say that it took me some time and I'm still unsure I got all the nuances.

This is a problem by itself: the rule tries to analyze a potentially large portion of the plan and then makes changes in positions that are very far from the entry point of the rule; it's hard to understand now, so it will be even harder to maintain in the long term, when the original intentions will be harder to recollect and when the language (and the planner) will be richer and more complex.

This said, I tried to understand if the rule is correct in all cases, and _my conclusion is that it's not_.
My understanding is that the intention is the following: if you have a plan with `sort | mv_expand | ... | sort`, you replace it with `mv_expand | sort | ... | sort`, in the assumption that eventually the first `sort` will be removed (because redundant?).
This is correct as long as there are no commands in between that drop possible candidate results, or for which the order actually matters.

The case that we are ignoring now is LIMIT: a query like `sort | mv_expand |  ... | limit | ... | sort` is different from `mv_expand | sort |  ... | limit | ... | sort`, especially when the first sort happens on the expanded key. Both `descendantMvExpand()` and `replaceChildUntilMvExpandAndOrderBy()` are not taking this case into consideration.

Cases we could miss tomorrow are all the commands that rely on the intermediate order, that include inline stats, commands that access previous records and other cases we can hardly completely define now.

In short, this rule would be much better defined as a local rule that converts `sort | mv_expand` -> `mv_expand | sort`, but unfortunately it's not generally valid, and defining all the cases when it's valid is risky and error prone. Also, the general usefulness of this change is not completely clear (to me at least).
We can try to fix the above, but we are paying a high cost to maintainability for arguably small advantages.

**Review by luigidellaquila — CHANGES_REQUESTED** [review]
Thanks @astefan, this is a complex topic and rationalizing it is not easy at all.
I left a comment about sort pushdown, that seems to miss some cases.
In general, I'd still prefer to keep it more simple, ie. push limit past mv_expand (keeping the limit inside mv_expand), plus meaningful error messages for cases that we cannot cover (ie, unbounded sort that still cannot be optimized)

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 5 above)
Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:489]: ++, agree that it's hard to verify this rule for all cases, esp. as we might add new operators that might invalidate it in the future, silently.

If we want to keep the `sort | mv_expand | ... | sort -> mv_expand | sort | ... | sort`, we probably should restrict the operators that can go in the ` ... ` part explicitly, ideally adding at least a unit test per such operator to verify that output doesn't change after optimization. That is a lot of work, though :/

[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:489]: > suggestion: a small example as javadoc would immensely help understand how this rule works.

My advice: take a look at the tests in LogicalPlanOptimizerTests and their javadoc, where the optimized plan is visible. Alternatively, `trace` logging on `org.elasticsearch.xpack.esql.optimizer` to see what each rule is doing when.

> Question: how is pushing MvExpand past an OrderBy beneficial?

In the description of the PR I have identified two cases for this type of pushdown. If the semantics of the query remains the same (and they are in those two situations I mentioned), it is beneficial to form a TopN by having sort (OrderBy) closer to a limit. Since every query now has a `limit`, the aim is to have the sort closer to it.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 6]
```diff
@@ -542,31 +517,97 @@ private static MvExpand descendantMvExpand(UnaryPlan unary) {
...
+            if (mvExpand != null) {
+                // if the filter is on the expanded values and the order of the results doesn't change, push down mv_expand past sort
+                if (isFiltered && hasOrderBy == false) {
+                    if (mvExpand.child() instanceof OrderBy) {
+                        return limit.replaceChild(replaceChildUntilMvExpandAndOrderBy((UnaryPlan) child, mvExpand));
+                    }
+                } else if (mvExpand.limit() == null) { // otherwise push down a copy of the limit before mv_expand
+                    var duplicateLimit = new Limit(limit.source(), limit.limit(), mvExpand.child());
+                    return limit.replaceChild(propagateDuplicateLimitUntilMvExpand(duplicateLimit, mvExpand, (UnaryPlan) child));
+                }
```
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:586]: No. Initially, I thought so too, but there is an edge case where there is still an infinite loop of removing the limit and adding back the limit. That edge case is in `LogicalPlanOptimizerTests.testInfinteLoopInLimitPushDown_WithLocalRelation`. `PropagateEmptyRelation` removes the duplicated `limit` and the mv_expand rule will add it again and again.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:489]: Thanks.

I still think it should be obvious what a rule does from either its name or a javadoc; the optimizer tests are not sufficient as they contain complete optimization runs, so there's no fast way to see what a rule does (in isolation). This rule is also sufficiently complex that skimming the code does not give an immediate idea of what to expect in terms of its behavior.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 7]
```diff
@@ -542,31 +517,97 @@ private static MvExpand descendantMvExpand(UnaryPlan unary) {
...
                     break;
                 }
             }
-            return null;
+
+            if (mvExpand != null) {
+                // if the filter is on the expanded values and the order of the results doesn't change, push down mv_expand past sort
+                if (isFiltered && hasOrderBy == false) {
+                    if (mvExpand.child() instanceof OrderBy) {
+                        return limit.replaceChild(replaceChildUntilMvExpandAndOrderBy((UnaryPlan) child, mvExpand));
```
[the remark this answers is not in the document]
Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:581]: We can leave it but I think it should be documented; this is definitely unexpected.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:489]: Wait, sorry - this is actually one of the simpler rules - I thought I left this comment further down.

Still - pushing down mv_expand past a sort is not an obvious optimization (obvious would be e.g. pushing down filters to reduce number of rows). A short comment explaining the reasoning for why it exists in the first place would be important IMHO.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:1432]: If you refer to topn(`salary`) yes, probably. This was my first attempt at mv_expand planning (an older, different PR). But `sort first_name` needs to go away as well because it has no effect on the results, nor on what happens after it.

Comment (alex-spies) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:1432]: Sorry, should have marked my comment more clearly: this is just a thought for future optimizations that I wanted to leave. (Out of scope of this PR)
